Step1 - 인수 테스트 기반 리팩터링

yulmucha opened the pull request:
안녕하세요.
리뷰 요청드립니다.
감사합니다.

Commit:
docs(README.md): add requirements

Commit:
refactor(LineService, Line, LineTest): move findUpStation(), getStations() to Line

Commit:
refactor(LineService, Line, LineTest): move the logic of adding a section to Line

Commit:
refactor(LineService, Line, LineTest): move the logic of removing a section to Line

Commit:
test(LineTest): add exceptional cases when adding a section

Commit:
refactor(LineService, StationService): remove a redundant method

Commit:
refactor(-Service, Line, Section, LineTest): throw the more specific exception instead of RuntimeException

Commit:
chore(LineService): add final keyword to fields and make findLineById() private

Commit:
refactor(Line, Sections, SectionsTest): move methods of Line to Sections

Commit:
refactor(Sections, SectionsTest): extract methods, and remove redundant if statements and streams

Commit:
chore(Sections): add empty check, and place private methods below public methods

Commit:
docs(READMD.md): mark an item as done

Commit:
chore(Line, LineService): make update() method more clear

enemfk777 commented on src/main/java/nextstep/subway/line/application/LineService.java:
```diff
@@ -12,28 +11,26 @@
+        Station downStation = stationService.findStationById(request.getDownStationId());
         Line persistLine = lineRepository.save(new Line(request.getName(), request.getColor(), upStation, downStation, request.getDistance()));
-        List<StationResponse> stations = getStations(persistLine).stream()
+        List<StationResponse> stations = persistLine.getStations().stream()
                 .map(it -> StationResponse.of(it))
                 .collect(Collectors.toList());
         return LineResponse.of(persistLine, stations);
```
이부분을 LineResponse에서 책임지고 만들어내는건 어떨까요?
service 내의 중복로직이 많이 제거될 것 같네요 :)

yulmucha replied:
결국 `Line`으로 만드는 거라 `LineResponse.of()` 파라미터에 `Line`만 남겨도 충분한 걸 놓쳤네요! 😅

enemfk777 commented on src/main/java/nextstep/subway/line/application/LineService.java:
```diff
@@ -43,30 +40,25 @@ public List<LineResponse> findLines() {
-    }
-
-
     public LineResponse findLineResponseById(Long id) {
```
조회용 메서드는 읽기전용 트랜잭션 설정을 해주는건 어떨까요?

yulmucha replied:
기본적으로 서비스에 `@Transactional(readOnly = true)`로 해 놓고, 쓰기 메서드에만 `@Transactional`을 붙이는 방식으로 수정했습니다.
조회만 할 때는 변경 감지를 위한 추가 작업을 하지 않아서 불필요한 메모리 사용을 방지할 수 있네요.

enemfk777 commented on src/main/java/nextstep/subway/line/application/LineService.java:
```diff
@@ -43,30 +40,25 @@ public List<LineResponse> findLines() {
     public void updateLine(Long id, LineRequest lineUpdateRequest) {
-        Line persistLine = lineRepository.findById(id).orElseThrow(RuntimeException::new);
-        persistLine.update(new Line(lineUpdateRequest.getName(), lineUpdateRequest.getColor()));
+        Line persistLine = lineRepository.findById(id).orElseThrow(NoSuchElementException::new);
```
private 메서드로 분리해주신 findLineById 를 사용할 수 있을것같아요 :)

enemfk777 commented on src/main/java/nextstep/subway/line/application/LineService.java:
```diff
@@ -77,106 +69,16 @@ public void addLineStation(Long lineId, SectionRequest request) {
-
-        return downStation;
+    private Line findLineById(Long id) {
+        return lineRepository.findById(id).orElseThrow(NoSuchElementException::new);
```
체이닝 메서드는 한줄로 쓰기보단 .마다 줄바꿈을 하면 가독성이 높아질 것 같아요 :)

yulmucha replied:
로직 옮기는 데만 급급했네요. 😅 동의하며 평소에는 실천하고 있습니다.

enemfk777 commented on src/main/java/nextstep/subway/line/domain/Sections.java:
```diff
@@ -0,0 +1,142 @@
+    @OneToMany(mappedBy = "line", cascade = {CascadeType.PERSIST, CascadeType.MERGE}, orphanRemoval = true)
+    private List<Section> elements = new ArrayList<>();
+
+    public Station lastUpStation() {
```
production 코드에서 sections 내부에서만 쓰이는 메서드여서 접근제한자가 public 일 이유가 없어보입니다 :)

yulmucha replied:
나중에라도 충분히 외부에서 사용될 만한 정보가 아닌가 싶어서 고민하다가 `public`으로 두었지만,
말씀을 듣고 보니, 실제로 사용될 때 `public`으로 바꿔도 될 건데 당장 읽는 사람에게 혼란만 주는 것 같네요.

yulmucha replied:
다만, 메서드가 테스트할 필요성이 있으면서 조회용이라 외부에서 사용해도 문제가 발생할 여지가 없다면 `public`이어도 괜찮지 않을까 싶습니다.

enemfk777 commented on src/main/java/nextstep/subway/line/domain/Sections.java:
```diff
@@ -0,0 +1,142 @@
+        }
+
+        Station upStation = elements.get(0).getUpStation();
+        while (upStation != null) {
+            Optional<Section> prevSection = findPrevSection(upStation);
+            if (!prevSection.isPresent()) {
+                break;
+            }
+            upStation = prevSection.get().getUpStation();
+        }
```
> 규칙 1: 한 메서드에 오직 한 단계의 들여쓰기(indent)만 한다.

객체지향 생활체조 원칙을 지켜보시는건 어떨까요? :)

enemfk777 commented on src/main/java/nextstep/subway/line/domain/Sections.java:
```diff
@@ -0,0 +1,142 @@
+        }
+
+        List<Station> stations = new ArrayList<>();
+        Station downStation = lastUpStation();
```
메서드명은 upstation인데 변수명이 downstation인건 언뜻 코드를 읽을 때 혼동을 줄 수 있을 것 같습니다

yulmucha replied:
동의합니다.
리팩터링 전의 이름인 `findUpStation`보다는 낫다고 생각했지만, 나중에 다시 읽을 때 한번은 저 스스로도 이해하는 데 시간이 좀 걸리더라고요.

yulmucha replied:
혼동이 없도록 첫 번째 `Section`을 찾도록 바꾸었습니다.
아, 방금 든 생각인데, 저 혼자서 마음대로 상행 쪽이 first고 하행 쪽이 last라고 생각한 게 아닌가 싶네요. `lastUpSection`으로 했어야 했나..

enemfk777 commented on src/main/java/nextstep/subway/line/domain/Sections.java:
```diff
@@ -0,0 +1,142 @@
+    }
+
+    private void validateStations(Station upStation, Station downStation) {
+        List<Station> stations = getStations();
+        boolean isUpStationExisted = stations.stream().anyMatch(it -> it == upStation);
+        boolean isDownStationExisted = stations.stream().anyMatch(it -> it == downStation);
```
sections에 station이 포함되는지만 확인하면 되는 로직으로 보이는데, 불필요한 정렬이들어가고 있네요 :)
section 스스로가 특정 station을 포함하고 있는지 반환할 수 있게 하면 어떨까요?

yulmucha replied:
앗 여기도 옮기는 데에만 정신이 팔렸었네요.

enemfk777 commented on src/main/java/nextstep/subway/line/domain/Sections.java:
```diff
@@ -0,0 +1,142 @@
+
+    private void rearrangeElementsFor(Section section) {
+        elements.stream()
+                .filter(it -> it.getUpStation() == section.getUpStation())
```
이부분도 section이 section을 파라미터로 전달받아 upstation이 같은 section인지, downstation이 같은 section인지 반환해볼 수 있을 것 같아요 :)

enemfk777 commented on src/main/java/nextstep/subway/line/domain/Sections.java:
```diff
@@ -0,0 +1,142 @@
+                .filter(it -> it.getUpStation() == section.getUpStation())
+                .findFirst()
+                .ifPresent(it -> {
+                    it.updateUpStation(section.getDownStation(), section.getDistance());
```
updateUpstation의 파라미터로 section을 전달 받는게 좀더 깔끔한 코드가 될 수 있을것 같습니다

yulmucha replied:
`Section`에 `updateUpStationAndDistanceFor(Section)` 메서드를 만들긴 했는데 `public`이라 좀 찜찜하네요.
`Sections`의 `rearrangeElementsFor()` 메서드 안에서는 조건에 따라 알맞게 사용되고 있지만, 이게 `Sections`와는 별개로 `Section`에 있는 `public` 메서드니까 충분히 잘못 사용될 수 있지 않나 싶어서요.
그래서 `updateUpStationAndDistanceFor()` 안에 검증하는 조건문 하나 더 넣어 보았습니다.
너무 과한 걱정일까요..

enemfk777 commented on src/main/java/nextstep/subway/member/application/MemberService.java:
```diff
@@ -21,12 +23,12 @@ public MemberResponse createMember(MemberRequest request) {
 
     public MemberResponse findMember(Long id) {
-        Member member = memberRepository.findById(id).orElseThrow(RuntimeException::new);
+        Member member = memberRepository.findById(id).orElseThrow(NoSuchElementException::new);
```
적절한 exception 설정 👍

enemfk777 commented on src/main/java/nextstep/subway/line/domain/Section.java:
```diff
@@ -56,17 +57,33 @@ public int getDistance() {
 
     public void updateUpStation(Station station, int newDistance) {
         if (this.distance <= newDistance) {
```
> 규칙 3: 모든 원시값과 문자열을 포장한다.

Distance는 0 보다 커야한다는 비즈니스 요구 사항이 있는 것 같은데 
객체로 포장해서 책임을 부여하는건 어떨까요?

yulmucha replied:
wrapper class `Distance`를 만들어 사용해 보았습니다

enemfk777 commented on src/main/java/nextstep/subway/line/domain/Sections.java:
```diff
@@ -0,0 +1,142 @@
+        Optional<Section> downSection = findSectionByUpStationSameAs(station);
+
+        if (downSection.isPresent() && upSection.isPresent()) {
+            Station newUpStation = upSection.get().getUpStation();
+            Station newDownStation = downSection.get().getDownStation();
+            int newDistance = downSection.get().getDistance() + upSection.get().getDistance();
```
두 section을 합치는게 sections의 책임일까요?
필요한 데이터는 모두 section이 가지고 있는데, section 스스로 책임질수 있게 리팩토링해보시는건 어떨까요?

yulmucha replied:
`Section`에 `merge()` 메서드를 만들었습니다.
처음엔 `static`으로 만들어서 두 `Section`을 받아야 하나 싶었어요.
하지만 그렇게 되면 비록 메서드는 `Section`에 존재하지만 `Section` object가 스스로 책임진다고 할 수 있을까.. 외부에서 getter로 데이터를 받아 조작하는 것과 비슷하지 않은가.. object 간의 상호 작용이 아닌 것 같다는 생각이 들어서 non-static으로 만들었습니다.

enemfk777 reviewed (changes requested):
안녕하세요 율님 :)
인수 테스트 기반 TDD 미션을 함께하게 된 리뷰어 강태엽입니다 🙇‍♂️

1단계 잘 수행해주셨네요 :)
몇가지 코멘트 드렸으니 한 번 슥 보시고 반영 or 의견 부탁드려요!!

그럼 이번 미션동안 잘부탁드려요 화이팅!! 💪

Commit:
refactor(LineService, LineResponse): from LineService to LineResponse, move the logic of mapping the list of StationResponse

Commit:
refactor(LineService): use @Transactional annotation in more specific way

Commit:
chore(LineService): use the existing function, and break some lines for better readability

Commit:
refactor(Sections, Section, SectionsTest): enhance readability, shift responsibility
- from Sections to Section, move the responsibility of comparing between Stations of Sections
- reduce indent depth
- rename functions and variables

Commit:
chore(Sections): relocate or rename methods

Commit:
refactor(Sections, Section): check more efficiently when validating the section to be added

Commit:
refactor(Sections, Section): shift responsibility from Sections to Section

Commit:
refactor(Distance, Section): use a wrapper class

enemfk777 commented on src/main/java/nextstep/subway/line/domain/Sections.java:
```diff
@@ -0,0 +1,144 @@
+    }
+
+    public List<Station> getStationsInOrder() {
+        if (elements.isEmpty()) {
```
elements가 비어있을 땐 exception을 발생시키면 어떨까요?
비즈니스 요구사항 상 elements가 비어있는 sections는 존재 할 수 없을 것 같아요

enemfk777 commented on src/main/java/nextstep/subway/line/domain/Sections.java:
```diff
@@ -0,0 +1,144 @@
+            throw new IllegalArgumentException("이미 등록된 구간 입니다.");
+        }
+
+        if (hasValidStation(section)) {
```
조건이 반대로 되어있는 것 같습니다 :)

yulmucha replied:
❗️이름을 반대로 지었네요.

enemfk777 commented on src/main/java/nextstep/subway/line/domain/Sections.java:
```diff
@@ -0,0 +1,144 @@
+                .filter(section -> section.hasSameUpStationAs(newSection))
+                .findFirst()
+                .ifPresent(section -> {
+                    section.updateUpStationAndDistanceFor(newSection);
+                    return;
```
람다로 표현해도 될 것 같아요 :)

enemfk777 commented on src/main/java/nextstep/subway/line/domain/Sections.java:
```diff
@@ -0,0 +1,144 @@
+    }
+
+    public void removeStation(Station station) {
+        if (elements.size() <= 1) {
```
매직넘버는 상수화 시켜보는건 어떨까요?

yulmucha replied:
이 경우에는 좀 애매하네요.
`size()`와 함께 `1` 자체가 그 의미를 충분히 나타내고 있어서 `ONE` 등으로 바꾸는 게 큰 의미가 없지 않을까 싶기도 하고..
`elements.size() > 1` 을 하나의 함수 `isPossibleToRemove()` 등으로 바꿀 순 있을 것 같습니다.

enemfk777 replied:
ONE이 아닌 MINIMUM_SECTION_SIZE 등의 비즈니스 요구사항을 나타낼수 있는 이름으로 명명하면 의미가 있을 것 같아요 :)

yulmucha replied:
그 생각을 못했네요 ㅎㅎ.. 감사합니다!!

enemfk777 commented on src/main/java/nextstep/subway/line/domain/Distance.java:
```diff
@@ -0,0 +1,58 @@
+        return value;
+    }
+
+    public void plus(Distance distanceNew) {
+        value += distanceNew.value;
```
Distance 연산에서 새로운 Distance 객체를 반환하도록 해서
기존 객체를 불변으로 만들어 보시는건 어떨까요?

enemfk777 reviewed (approved):
안녕하세요 율님 :)

많은 피드백이었는데 잘 반영해 주셨네요!!
몇 가지 고민해 볼 사항을 추가로 코멘트 드렸으니 확인 부탁드립니다!

그럼 다음 단계 진행을 위해 이만 머지하겠습니다!!
